conductor/lib/uploader.py

```diff
@@ -295,6 +371,83 @@ def do_upload(self, upload_url, filename, md5):
                                                 tries=1,
                                                 use_api_key=True)
 
+    def do_multipart_upload(self, upload, filename, md5):
+        """
+        Files will be split into partSize returned by the FileAPI and hydrated
+        once all parts are uploaded. On successful part upload, response headers will contain an ETag.
+        This value must be tracked along with the part number in order to complete and hydrate the file.
+        """
+        uploads = []
+        complete_payload = {
+            "uploadID": upload["uploadID"],
+            "hash": md5,
+            "completedParts": [],
+            "project": self.project
+        }
+
+        # iterate over parts and upload
+        for part in upload["parts"]:
+            resp_headers = self._do_multipart_upload(
+                upload_url=part["url"],
+                filename=filename,
+                part_number=part["partNumber"],
+                part_size=upload["partSize"]
+            )
+
+            if resp_headers:
+                uploads.append(upload["uploadID"])
+                completed_part = {
+                    "partNumber": part["partNumber"],
+                    "etag": resp_headers['ETag'].strip('"')
+                }
+                complete_payload["completedParts"].append(completed_part)
+
+        # Complete multipart upload in order to hydrate file for availability
+        uri_path = '/api/v2/files/multipart/complete'
+        headers = {'Content-Type': 'application/json'}
+        self.api_client.make_request(uri_path=uri_path,
+                                     verb='POST',
+                                     headers=headers,
+                                     data=json.dumps(complete_payload),
+                                     raise_on_error=True,
+                                     use_api_key=True)
+
+        return uploads
+
+    @common.DecRetry(retry_exceptions=api_client.CONNECTION_EXCEPTIONS, tries=5)
+    def _do_multipart_upload(self, upload_url, filename, part_number, part_size):
+        with open(filename, 'rb') as fh:
+            # seek to the correct part position
+            start = (part_number - 1) * part_size
+            fh.seek(start)
```

Comment:
Regarding https://github.com/AtomicConductor/conductor_client/pull/290/files#r435959647, 
yeah, i'm wondering about the part size logic/math. Though efficient/minimal, I'm wondering if we should be more explicit.   
For example, the code in this PR is currently deriving which part of the file to seek to by calculating the partSize by the partNumber it's been given (minus 1).   This is an efficient, and elegently minimal implementation, but it's relying on a couple implementation decisions.

In contrast, I'm wondering if we should follow a more descriptive approach that more explicitly (yet generically) describes to the client what is needed.

So instead of:

```python
"multiPartURLs": [
    {
        ...
        "partSize": 1073741824,
        "parts: [
            {
                "partNumber": 1,
                "url: "https://www.signedurlexample.com/signature1"
            },
            {
                "partNumber": 2,
                "url: "https://www.signedurlexample.com/signature1"
            }
        ]
    }
]
```

We use:
```python
"multiPartURLs": [
    {
        ...
        "size": 2147483648,  # the size of the entire file  (replaces the partSize field)
        "parts: [
            {
                "partNumber": 1,
                "url: "https://www.signedurlexample.com/signature1",
                "range": "0-1073741823"
            },
            {
                "partNumber": 2,
                "url: "https://www.signedurlexample.com/signature1",
                "range": "1073741824-2147483648"
            }
        ]
    }
]
```



There are a few benefits to this:

1. The explicit range makes it clear as to what to upload.  There's no need to derive this information based on the `partNumber` and `partSize`.

2. it inherently provides the size of each part.

3. it remedies the misleading situation where the "partSize" doesn't actually reflect the size of the part that is being uploaded (which is oftentimes the case for the last Part of a multipart).

4. By designating an explicit range for each Part, it opens up the possibility of having dynamic part sizes (i.e. the backend can potentially adjust part sizes, rather than always targeting 1GB, etc).

5. (arguably/superficially) Using a range for each part more closely aligns with a traditional http convention (i.e. using a Range header).

Reply:
Personally I don't see many major reasons to switch to the latter, but since this is a new handler we probably want to get this right/agreed upon before regretting it later on.

Point 3, technically if someone were to make a different implementation of the uploader and received a multipart upload response, there is nothing stopping them from NOT using all the preSignedURLs. For example a 6GB file, current implementation would create and respond with 6 preSignedURLs but they could use only 2 (as long as they are all under 5GB and more than 5MB except for the last) of them if they want a 5GB + 1GB, send a completeMultiPart with those ETags and all would be well.

BackBlaze for example calls it a [recommendedPartSize](https://www.backblaze.com/b2/docs/b2_authorize_account.html) when dealing with large uploads would that being more preferable/less misleading?
 
Point 4 , is interesting but that could always just be a calculated value on the backend there is not stopping us from making the `partSize` completely dynamic based on fileSize or other considerations.

These are mostly just my personal opinions, if you and @flebel feel strongly about changing it to that I have no problem making the necessary changes on the file-api side to reflect that.

Reply:
If the current implementation feels good to you, carry on